extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java

```diff
@@ -377,4 +295,110 @@ private static RowSignature computeSignatureForFinalWindowStage(RowSignature row
         finalWindowClusterBy.getColumns()
     );
   }
+
+  /**
+   * Represents a window stage in a query execution.
+   * Each stage can contain a sort operator, a partition operator, and multiple window operators.
+   */
+  private static class WindowStage
+  {
+    private NaiveSortOperatorFactory sortOperatorFactory;
+    private NaivePartitioningOperatorFactory partitioningOperatorFactory;
+    private final List<WindowOperatorFactory> windowOperatorFactories;
+
+    public WindowStage()
+    {
+      this.windowOperatorFactories = new ArrayList<>();
+    }
+
+    public void addOperatorFactory(OperatorFactory op)
+    {
+      if (op instanceof NaiveSortOperatorFactory) {
+        this.sortOperatorFactory = (NaiveSortOperatorFactory) op;
+      } else if (op instanceof NaivePartitioningOperatorFactory) {
+        this.partitioningOperatorFactory = (NaivePartitioningOperatorFactory) op;
+      } else {
+        this.windowOperatorFactories.add((WindowOperatorFactory) op);
+      }
+    }
+
+    public List<OperatorFactory> getOperatorFactories()
+    {
+      List<OperatorFactory> operatorFactories = new ArrayList<>();
+      if (sortOperatorFactory != null) {
+        operatorFactories.add(sortOperatorFactory);
+      }
+      if (partitioningOperatorFactory != null) {
+        operatorFactories.add(partitioningOperatorFactory);
+      }
+      operatorFactories.addAll(windowOperatorFactories);
+      return operatorFactories;
+    }
+
+    public List<String> getPartitionColumns()
+    {
+      return partitioningOperatorFactory != null ? partitioningOperatorFactory.getPartitionColumns() : new ArrayList<>();
+    }
+
+    public List<WindowOperatorFactory> getWindowOperatorFactories()
+    {
+      return windowOperatorFactories;
+    }
+
+    public ShuffleSpec findShuffleSpec(int partitionCount)
```

Comment:
Can you elaborate on how it should be decided here?

Also, for this PR, I'm trying to keep this as a pure refactoring PR, so would like to avoid making any functional changes if that's okay.